### PR TITLE
8283075: Bad IllegalArgumentException message for out of range rank from ClassDesc.arrayType(int)

### PR DESCRIPTION
**Pull request by jddarcy**
Improving the exception messages for out-of-supported-range array types.

I'll update copyrights before pushing.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8283075](https://bugs.openjdk.java.net/browse/JDK-8283075): Bad IllegalArgumentException message for out of range rank from ClassDesc.arrayType(int)

### Reviewers
 * [Vicente Romero](https://openjdk.java.net/census#vromero) (@vicente-romero-oracle - **Reviewer**) ⚠️ Review applies to 1723d7106f8e28afc15430cbfe6d0732d6ca50ba
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to 1723d7106f8e28afc15430cbfe6d0732d6ca50ba

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7812/head:pull/7812` \
`$ git checkout pull/7812`

Update a local copy of the PR: \
`$ git checkout pull/7812` \
`$ git pull https://git.openjdk.java.net/jdk pull/7812/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7812`

View PR using the GUI difftool: \
`$ git pr show -t 7812`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7812.diff">https://git.openjdk.java.net/jdk/pull/7812.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/lang/constant/ClassDesc.java [hunk 1]
```diff
@@ -167,14 +168,23 @@ default ClassDesc arrayType() {
...
-            throw new IllegalArgumentException("rank: " + currentDepth + rank);
+        int netRank;
+        try {
+            netRank = Math.addExact(currentDepth, rank);
+        } catch (ArithmeticException ae) {
+            throw new IllegalArgumentException("Integer overflow in rank computation");
+        }
+        if (rank <= 0 || netRank > ConstantUtils.MAX_ARRAY_TYPE_DESC_DIMENSIONS)
+            throw new IllegalArgumentException("rank: " + netRank +
+                                               " exceeds maximum supported dimension of " +
```
Comment (vicente-romero-oracle) [inline src/java.base/share/classes/java/lang/constant/ClassDesc.java:186]: if the `rank < 0` it won't be exceeding the maximum supported dimensions so this message only applies to the second condition

**Review by vicente-romero-oracle — APPROVED** [review]
lgtm

**Review by AlanBateman — APPROVED** [review]

File: src/java.base/share/classes/java/lang/constant/ClassDesc.java [hunk 2]
```diff
@@ -167,14 +168,27 @@ default ClassDesc arrayType() {
...
      * greater than 255
      * @jvms 4.4.1 The CONSTANT_Class_info Structure
      */
     default ClassDesc arrayType(int rank) {
-        int currentDepth = ConstantUtils.arrayDepth(descriptorString());
-        if (rank <= 0 || currentDepth + rank > ConstantUtils.MAX_ARRAY_TYPE_DESC_DIMENSIONS)
-            throw new IllegalArgumentException("rank: " + currentDepth + rank);
+        int netRank;
+        if (rank <= 0) {
+            throw new IllegalArgumentException("rank " + rank + "is not a positive value");
```
Comment (plevart) [inline src/java.base/share/classes/java/lang/constant/ClassDesc.java:179]: There's a space missing in the IAE message after the rank number. For example, if rank == 0, the message would be: "rank 0is not positive value" ...

**Review by plevart — CHANGES_REQUESTED** [review]

File: src/java.base/share/classes/java/lang/constant/ClassDesc.java [hunk 3]
```diff
@@ -167,14 +168,27 @@ default ClassDesc arrayType() {
...
+        }
+        try {
+            int currentDepth = ConstantUtils.arrayDepth(descriptorString());
+            netRank = Math.addExact(currentDepth, rank);
+            if (netRank > ConstantUtils.MAX_ARRAY_TYPE_DESC_DIMENSIONS) {
+                throw new IllegalArgumentException("rank: " + netRank +
```
Comment (ExE-Boss) [inline src/java.base/share/classes/java/lang/constant/ClassDesc.java:185]: This doesn’t need to use <code>[Math.addExact]\(<var>currentDepth</var>, <var>rank</var>\)</code>, as <code><var>currentDepth</var></code> will always be in [0, 255], and <code><var>rank</var></code> here is confined to [1, 2<sup>31</sup>-1], so it’s possible to check for <code><var>netRank</var> &lt; 0</code> and use <code>[Integer.toUnsignedString]\(<var>netRank</var>\)</code> (or <code>[Integer.toUnsignedLong]\(<var>netRank</var>\)</code> and let <code>[StringConcatFactory]</code> deal with it):

```suggestion
            netRank = currentDepth + rank;
            if (netRank > ConstantUtils.MAX_ARRAY_TYPE_DESC_DIMENSIONS || netRank < 0) {
                throw new IllegalArgumentException("rank: " + Integer.toUnsignedLong(netRank) +
```

[Integer.toUnsignedLong]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Integer.html#toUnsignedLong(int)
[Integer.toUnsignedString]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Integer.html#toUnsignedString(int)
[Math.addExact]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Math.html#addExact(int,int)
[StringConcatFactory]: https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/invoke/StringConcatFactory.html

File: src/java.base/share/classes/java/lang/constant/ClassDesc.java (hunk 2 above)
Comment (jddarcy) [reply src/java.base/share/classes/java/lang/constant/ClassDesc.java:179]: Good catch; will fix before pushing.

File: src/java.base/share/classes/java/lang/constant/ClassDesc.java (hunk 3 above)
Comment (liach) [reply src/java.base/share/classes/java/lang/constant/ClassDesc.java:185]: Notice the overflow case is already handled by the `ArithmeticException` caught in the larger try block.

**Pull request closed without merging**